[datetime] fix(DateInput): Improve blur handling to avoid unexpected popover closures

ejose19 opened the pull request:
#### Fixes #3603

#### Changes proposed in this pull request:

The poposed solution is to avoid the case where minDate/maxDate is set, and using the change month arrows makes the popover to close on chrome browser, upon debugging, the issue is that document.activeElement is returning "body" instead of the change month button, which makes ```handlePopoverBlur``` set up execute incorrectly and subsequently calling ```handleClosePopover``` which causes the described issue.

#### Reviewers should focus on:

While this solution works correctly for that case, it makes "Popover closes when first day of the month is blurred" test fail, and after evaluating both use cases, it seems the min/max date has more weight, and so losing that "feature" is worth it.

Here's the [sandbox](https://codesandbox.io/s/old-butterfly-spcxx) (open in chrome) and without selecting any date, move 2 months before/after and notice how the popover closes incorrectly.

adidahiya reviewed (changes requested):
see my comment https://github.com/palantir/blueprint/issues/3603#issuecomment-689126068

ejose19 commented:
@adidahiya I think by the point we reach ```handlePopoverBlur``` it's too late to do something already. However at ```registerPopoverBlurHandler``` we can pass an argument to check if ```onMonthChange``` called this function, and which arrow triggered the event (left or right). Then since ```lastTabbableElement``` will be undefined, if the flag is set we assign a better candidate for ```lastTabbableElement```.

However what happens if theoretically a month change leads the popover to a state without any tabbable element? Besides from that let me know what you think about this idea.

ejose19 committed:
fix: dateInput popover closure when navigating months

ejose19 committed:
refactor: improve blur management on dateInput

ejose19 commented:
@adidahiya I've refactored this PR, now it should cover all use cases and even improve existing function-ability which had another unreported bug.

To reproduce the other bug, on a DateInput will nul value, place the active element on the first day of the month, and then press left arrow key, notice how the popover closes. This is because we're making ```handlePopoverBlur``` fire when DayPicker Day elements lose blur, but this can be a normal operation if the user is just navigating using the arrow buttons.

I also removed the step where it makes ```lastTabbableElement``` the document.activeElement, which is incorrect. An example of this mistake is having the dateinput value null, the time element being shown, and navigating using the keyboard arrows, when changing months it will report the active DayPicker Day element as ```lastTabbableElement```, but in practice it's still the milliseconds input.

adidahiya commented on packages/datetime/src/dateInput.tsx:
```diff
@@ -205,8 +205,16 @@ export class DateInput extends AbstractPureComponent2<IDateInputProps, IDateInpu
+            // If the user tabs on a DayPicker Day element and the lastElementInPopover is also a DayPicker Day
+            // element, the popover should be closed
+            onDayKeyDown: (day, modifiers, e) => {
+                if (e.key === "Tab" && !e.shiftKey && this.lastElementInPopover.classList.contains("DayPicker-Day")) {
```
```suggestion
                if (e.key === "Tab" && !e.shiftKey && this.lastElementInPopover?.classList.contains(Classes.DATEPICKER_DAY)) {
```

adidahiya commented on packages/datetime/src/dateInput.tsx:
```diff
@@ -205,8 +205,16 @@ export class DateInput extends AbstractPureComponent2<IDateInputProps, IDateInpu
+        const dayPickerProps: DayPickerProps = {
             ...this.props.dayPickerProps,
+            // If the user tabs on a DayPicker Day element and the lastElementInPopover is also a DayPicker Day
+            // element, the popover should be closed
```
I don't quite follow this comment, why is this the case?

ejose19 replied:
From your https://github.com/palantir/blueprint/issues/3603#issuecomment-689126068
> but I would still like to retain the functionality added in #2093 which closes the DateInput popover after a user TABs out of the last tabbable element in the popover.

Which is basically a remark with that comment, if you feel it's redundant I can remove it.

adidahiya replied:
but what about "...and the lastElementInPopover is also a DayPicker Day element"? why is this guard necessary? the comment leaves me wondering what the other possibilities are (I assume it refers to cases where a time picker is rendered, making that the last tabbable element?)

ejose19 replied:
Yes, I tried all combinations and that extra guards need to be in place for when there's any tabbable element below the calendar (like time picker, but it would cover any new element added in the future)

adidahiya replied:
ok, then that begs the question... which code is responsible for closing the popover when there is a tabbable element (like TimePicker) below the calendar? your implementation seems to work fine, but I want to make sure the code comments are clear and comprehensible for the next person who reads this code. I'm pulling your branch now to figure this out for myself

ejose19 replied:
```registerPopoverBlurHandler``` along ```handlePopoverBlur``` is handling that (like originally implemented), just excluding the 2 cases mentioned in the comment. It's actually "agnostic" regarding if the last tabbable element is below the calendar or not, that state variable holds the last tabbable element inside the popover and listens for blur events.

Actually, now that you mention it, I found that this code https://github.com/palantir/blueprint/blob/408a8a73c4b49f1642abadcd0cde3e113d82e055/packages/datetime/src/dateInput.tsx#L446-L447 is also incorrect. They are storing the "last tabbed element", not "last tabbable element" which will cause an issue if the user shift+tab and then tab. I will fix that part to use the same formula as ```registerPopoverBlurHandler``` (also maybe a documentation to ```lastElementInPopover``` meaning "last element in popover that is tabbable, and the one that triggers closure if the user press tabs on it"

adidahiya commented on packages/datetime/src/dateInput.tsx:
```diff
@@ -413,12 +421,23 @@ export class DateInput extends AbstractPureComponent2<IDateInputProps, IDateInpu
+            // and would close the Popover unexpectedly
+            // - On disabled change months buttons
+            // - DayPicker day elements, their "blur" will be managed at its own onKeyDown
+            // event listener for improving UX
```
what does this comment line refer to? it feels a little out of place

ejose19 replied:
Actually one needs to read the entire comment to understand it.

```ts
// Exclude the following blur operations that makes "body" the activeElement
// and would close the Popover unexpectedly
// - On disabled change months buttons
// - DayPicker day elements, their "blur" will be managed at its own onKeyDown
// event listener for improving UX
const isChangeMonthEvt = eventTarget.classList.contains(Classes.DATEPICKER_NAV_BUTTON);
const isChangeMonthButtonDisabled = isChangeMonthEvt && (eventTarget as HTMLButtonElement).disabled;
const isDayPickerDayEvt = eventTarget.classList.contains(Classes.DATEPICKER_DAY);
if (!isChangeMonthButtonDisabled && !isDayPickerDayEvt) {
    this.handleClosePopover();
}
```

So the excluded blur events are ```isChangeMonthButtonDisabled``` and ```isDayPickerDayEvt``` as you can see in the if condition. Any idea to make it more clear?

adidahiya replied:
oh, sorry, I wasn't reading the last line as a continuation of the 2nd bullet. in that case just remove these words, `onKeyDown` is enough info:

```suggestion
```

adidahiya commented on packages/datetime/src/dateInput.tsx:
```diff
@@ -413,12 +421,23 @@ export class DateInput extends AbstractPureComponent2<IDateInputProps, IDateInpu
+            // - On disabled change months buttons
+            // - DayPicker day elements, their "blur" will be managed at its own onKeyDown
+            // event listener for improving UX
+            const isChangeMonthEvt = eventTarget.classList.contains("DayPicker-NavButton");
```
```suggestion
            const isChangeMonthEvt = eventTarget.classList.contains(Classes.DATEPICKER_NAVBUTTON);
```

you would need to add this class to `common/classes.ts` in this package

adidahiya commented on packages/datetime/src/dateInput.tsx:
```diff
@@ -413,12 +421,23 @@ export class DateInput extends AbstractPureComponent2<IDateInputProps, IDateInpu
+            // event listener for improving UX
+            const isChangeMonthEvt = eventTarget.classList.contains("DayPicker-NavButton");
+            const isChangeMonthButtonDisabled = isChangeMonthEvt && (eventTarget as HTMLButtonElement).disabled;
+            const isDayPickerDayEvt = eventTarget.classList.contains("DayPicker-Day");
```
```suggestion
            const isDayPickerDayEvt = eventTarget.classList.contains(Classes.DATEPICKER_DAY);
```

adidahiya reviewed (changes requested):
[docs preview](https://35010-71939872-gh.circle-artifacts.com/0/packages/docs-app/dist/index.html#datetime/dateinput)

this does appear to fix the issue, so the UX changes are ok

ejose19 committed:
refactor: apply PR recommendations

adidahiya commented on packages/datetime/src/common/classes.ts:
```diff
@@ -36,6 +36,7 @@ export const DATEPICKER_FOOTER = `${DATEPICKER}-footer`;
 export const DATEPICKER_MONTH_SELECT = `${DATEPICKER}-month-select`;
 export const DATEPICKER_YEAR_SELECT = `${DATEPICKER}-year-select`;
 export const DATEPICKER_NAVBAR = `${DATEPICKER}-navbar`;
+export const DATEPICKER_NAVBUTTON = `${DATEPICKER}-NavButton`;
```
actually it's 
```suggestion
export const DATEPICKER_NAVBUTTON = `DayPicker-NavButton`;
```

sorry, I know the "date" vs "day" thing is confusing in this file

adidahiya commented on packages/datetime/src/dateInput.tsx:
```diff
@@ -205,8 +205,20 @@ export class DateInput extends AbstractPureComponent2<IDateInputProps, IDateInpu
-        const dayPickerProps = {
+        const dayPickerProps: DayPickerProps = {
             ...this.props.dayPickerProps,
+            // If the user tabs on a DayPicker Day element and the lastElementInPopover is also a DayPicker Day
```
I think this could be a little more clear, making it explicit that it's TAB being pressed while the day is focussed, not the TAB moving focus _onto_ the day element

```suggestion
            // If the user presses the TAB key on a DayPicker Day element and the lastElementInPopover is also a DayPicker Day
```

ejose19 committed:
Apply suggestions from code review

Co-authored-by: Adi Dahiya <adi.dahiya14@gmail.com>

ejose19 committed:
fix: handlePopoverBlur when relatedTarget != null

adidahiya commented:
[docs preview build](https://35250-71939872-gh.circle-artifacts.com/0/packages/docs-app/dist/index.html#datetime/dateinput)

confirmed in Chrome, FF, IE11 👍 

thanks @ejose19
